irinterp: improve semi-concrete interpretation accuracy

aviatesk opened the pull request:
Currently semi-concrete interpretation can end up with more inaccurate result than usual abstract interpretation because `src.ssavaluetypes` are all widened when cached so semi-concrete interpretation can't use extended lattice information of `SSAValue`s. This commit tries to fix it by making `CodeInstance` keep extended lattice information of `SSAValue`s that are widened and allowing semi-concrete interpretation to override `src.ssavaluetypes` when it uncompressed `src` from `CodeInstance`.

I found there are other chances when semi-concrete interpretation can end up with inaccurate results, but that is unrelated to this and should be fixed separately.

aviatesk committed:
improve semi-concrete interpretation accuracy, take 1

Currently semi-concrete interpretation can end up with more inaccurate
result than usual abstract interpretation because `src.ssavaluetypes`
are all widened when cached so semi-concrete interpretation can't use
extended lattice information of `SSAValue`s. This commit tries to fix it
by making `CodeInstance` keep extended lattice information of `SSAValue`s
that are widened and allowing semi-concrete interpretation to override
`src.ssavaluetypes` when it uncompressed `src` from `CodeInstance`.

I found there are other chances when semi-concrete interpretation can
end up with inaccurate results, but that is unrelated to this and should
be fixed separately.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

Keno commented:
Not really a huge fan of this approach. Can we just stop widening there? That's done for the benefit of codegen, but we've moved both `Const` and `PartialStruct` into Core, so codegen could definitely look through those to get to its type information.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47994#issuecomment-1364656269) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/d1dfeb6_vs_21529a9/report.md).

aviatesk committed:
Revert "improve semi-concrete interpretation accuracy"

This reverts commit 67d9dc57709b37c8fe17d817c109efd8e2c8188c.

aviatesk committed:
improve semi-concrete interpretation accuracy, take 2

Currently semi-concrete interpretation can end up with more inaccurate
result than usual abstract interpretation because `src.ssavaluetypes`
are all widened when cached so semi-concrete interpretation can't use
extended lattice information of `SSAValue`s. This commit tries to fix it
by making the codegen system recognize extended lattice information such
as `Const` and `PartialStruct` and not-widening those information from
`src::CodeInfo` when caching.

I found there are other chances when semi-concrete interpretation can
end up with inaccurate results, but that is unrelated to this and should
be fixed separately.

aviatesk commented:
> Can we just stop widening there?

es, I just implemented your suggestion. We may want to switch back to my first approach once our system becomes able to run code produced by an external `AbstractInterpreter`, which may contain arbitrary lattice elements. However, at the current situation, your approach seems cleaner.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47994#issuecomment-1364763255) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/c769337_vs_162ee48/report.md).

aviatesk commented on src/interpreter.c:
```diff
@@ -210,7 +210,7 @@ static jl_value_t *eval_value(jl_value_t *e, interpreter_state *s)
 #ifndef JL_NDEBUG
         JL_GC_PUSH1(&val);
-        jl_typeassert(val, jl_fieldref_noalloc(e, 1));
+        jl_typeassert(val, jl_widen_core_extended_info(jl_fieldref_noalloc(e, 1)));
```
`analyzegc` complains on this line. Could someone tell me how I can avoid it?

Keno replied:
I didn't look at the error message, but I assume the issue is that `jl_widen_core_extended_info` can return a freshly-allocated result in the `Type` case, which isn't rooted anywhere. You can fix that by just changing the `JL_GC_PUSH1` to `JL_GC_PUSH2` above (on some other variable `jl_value_t *typ = NULL` (note the value at PUSH time matters, it needs to always be a valid pointer or `null` between the push and the pop)). Then just assign the result of the widening to that variable and it'll be rooted for the call.

aviatesk replied:
Thanks so much for your help! That seems to work.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47994#issuecomment-1365839653) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/252a47d_vs_a9e0545/report.md).

vtjnash commented:
We primarily used to do this to avoid exploding the precompile caches with garbage data that wastes space and costs GC time permanently. Are we sure this PR has not introduced a regression in that area, since we are recently starting to see test failures from an excess of roots being created (e.g. https://github.com/JuliaLang/julia/pull/48089)?

aviatesk commented:
I haven't checked anything in that area yet. Maybe I should check images size difference first, or is there any other recommended way to check the regression?
